packages/styletron-react/src/index.js

```diff
@@ -107,6 +114,52 @@ type createStyledOpts = {
   ) => React.ComponentType<any>,
 };
 
+function checkNoopEngine(engine: StandardEngine) {
+  // if no engine provided, we default to no-op, handy for tests
+  // however, print a warning in other envs
+  if (process.env.NODE_ENV !== "test") {
+    engine.noopEngine &&
+      // eslint-disable-next-line no-console
+      console.warn(
+        __DEV__
+          ? `
+Styletron has been switched to a no-op (test) mode.
+
+A Styletron styled component was rendered, but no Styletron engine instance was provided in React context.
+
+Did you forget to provide a Styletron engine instance to React context via using the Styletron provider component?
+
+Note: Providers and Consumers must come from the exact same React.createContext call to work.
+If your app has multiple instances of the "styletron-react" package in your node_module tree,
+your Provider may be coming from a different React.createContext call, which means the styled components
+will not recieve the provided engine instance. This scenario can arise, for example, when using "npm link".
+`
+          : `Styletron Provider is not set up. Defaulting to no-op.`,
+      );
+  }
+}
+
+export function useStyletron() {
+  const styletronEngine: StandardEngine = React.useContext(StyletronContext);
+  const debugEngine = React.useContext(DebugEngineContext);
+  const hydrating = React.useContext(HydrationContext);
+  return [
+    function css(style: StyleObject) {
+      checkNoopEngine(styletronEngine);
+      const className = driver(style, styletronEngine);
+      if (__BROWSER__ && __DEV__ && debugEngine && !hydrating) {
+        const {stack, message} = new Error("stacktrace source");
+        const debugClassName = debugEngine.debug({
```

Comment:
I'm wondering if this could be put into `useMemo` so this is only called once (rather than for each `css` call. This might save some unnecessary work.

Reply:
I played with it a bit but no luck. The problem is that it get memoized when `hydrating` or if I remove that from the guarding condition `if (__BROWSER__ && __DEV__ && debugEngine && !hydrating) {}`, then I get the mismatch server/client error. HMR just works (with or w/o the memo). It even fixes the last sourceMap link problem and starts pointing to the right place :)

<img width="384" alt="Screen Shot 2019-04-05 at 3 46 34 PM" src="https://user-images.githubusercontent.com/1387913/55660216-0e7b3580-57ba-11e9-9fd6-bbd79c2f065e.png">

Anyway, here's what I ended up with before I gave up:

```jsx
export function useStyletron() {
  const styletronEngine: StandardEngine = React.useContext(StyletronContext);
  const debugEngine = React.useContext(DebugEngineContext);
  const hydrating = React.useContext(HydrationContext);
  checkNoopEngine(styletronEngine);
  return [
    function css(style: StyleObject) {
      const className = driver(style, styletronEngine);
      const debugClassName =
        React.useMemo(
          () => {
            if (__BROWSER__ && __DEV__ && debugEngine) {
              const {stack, message} = new Error("stacktrace source");
              return debugEngine.debug({
                stackInfo: {stack, message},
                stackIndex: 5,
              });
            }
            return "";
          },
          [className],
        ) || "";
      if (debugClassName !== "") {
        return `${debugClassName} ${className}`;
      }
      return className;
    },
  ];
}
```

**It has two problems:**
- It also doesn't add the `debug` className on the initial render (not sure why)
- server/client mismatch warning

Also, not sure if `className` makes completely sense as useMemo's dep in this case. Note that the guarding condition needs to be inside of the memo hook, otherwise React explodes.

Reply:
What about something like:

```js
export function useStyletron() {
  const styletronEngine: StandardEngine = React.useContext(StyletronContext);
  const debugEngine = React.useContext(DebugEngineContext);
  const hydrating = React.useContext(HydrationContext);
  checkNoopEngine(styletronEngine);
  return [
    function css(style: StyleObject) {
      const className = driver(style, styletronEngine);
      // Use ref for error to avoid having stack trace inside of hooks
      const errorRef = useRef(new Error("stacktrace source"));
      const debugClassName = useMemo(
        () => {
          if (__BROWSER__ && __DEV__ && debugEngine && !hydrating) {
            const { stack, message } = errorRef.current;
            return debugEngine.debug({
              stackInfo: { stack, message },
              stackIndex: 1
            });
          }
          return "";
        },
        [hydrating]
      );
      return debugClassName ? `${debugClassName} ${className}` : className;
    }
  ];
}
```

Reply:
That works great! Thanks! The PR is updated.